[VedJoshi] iP

VedJoshi opened the pull request:
# Agave Chatbot

Agave is a chatbot that helps users manage tasks efficiently through a simple text-based interface. It allows you to create, track, and manage tasks in a straightforward way.

## Features:

- **Fast** and efficient task management
- *Easy* to use
- Runs locally on your machine

### How to use:
1. Download the chatbot [here](https://github.com/VedJoshi/ip).
2. Run the application.
3. Add tasks using commands.
4. Mark tasks as done or delete them when completed.

### Example Code:
```java
public class Task {
    private String description;
    private boolean isDone;
    
    public Task(String description) {
        this.description = description;
        this.isDone = false;
    }
}
```

## Example Output
### Task List:
- [x] Todo read a book
- [ ] Deadline read a book by the 12th of September 2024

Agave simplifies task management so you can focus on what really matters. Get started today! ✨

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

VedJoshi committed:
Level 0, new name (Agave), greeting and exit

VedJoshi committed:
Add UserInterface class, handle UI + I/O

VedJoshi committed:
Add run method as driver code, implemented testing

VedJoshi committed:
New input lines for testing

VedJoshi committed:
New task management system, level 2

VedJoshi committed:
Merge pull request #1 from VedJoshi/task_1.0

New task management system, level 2

VedJoshi committed:
Revert "New task management system, level 2"

VedJoshi committed:
Merge pull request #2 from VedJoshi/revert-1-task_1.0

Revert "New task management system, level 2"

VedJoshi committed:
Add level 2 task requirements

VedJoshi committed:
Add mark/unmark feature

VedJoshi committed:
Merge pull request #3 from VedJoshi/Level_3

Add mark/unmark feature

VedJoshi committed:
New task types minimum input validation

VedJoshi committed:
Merge pull request #4 from VedJoshi/Level_4

New task types minimum input validation

VedJoshi committed:
Level 5 reqs, a bit of pokemon

VedJoshi committed:
Merge pull request #5 from VedJoshi/Level_5

lvl 5 stuff, only handle user input checks.

VedJoshi committed:
Changed structure, main run() method easier to read

VedJoshi committed:
New inputs for ui-text testing, all passed

VedJoshi committed:
Merge pull request #6 from VedJoshi/A-TextUiTesting

New inputs for ui-text testing, all passed

VedJoshi committed:
Updated to use array list, added remove task feature

VedJoshi committed:
Merge pull request #7 from VedJoshi/Task_6

Updated to use array list, added remove task feature

VedJoshi committed:
Using buffered reader/writer for lvl 7 task

VedJoshi committed:
File reading and writing methods

VedJoshi committed:
To fix: getCorrectFormat(), out of bounds error, string formatting

VedJoshi committed:
Fixed marking issue

VedJoshi committed:
Fixed parsing issue

VedJoshi committed:
Merge pull request #8 from VedJoshi/Level_7

Level 7

VedJoshi committed:
Added comment for ToDo class

VedJoshi committed:
Merged branch lvl 7 into master

VedJoshi committed:
Added time object functionality

VedJoshi committed:
Merge branch lvl8 into master

VedJoshi committed:
Changed error message to adhere to expected input

VedJoshi committed:
Refactor start

VedJoshi committed:
Add OOP + required classes

VedJoshi committed:
Merge branch 'More_OOP'

VedJoshi committed:
Package successfully

VedJoshi committed:
Merge branch 'Packaging'

VedJoshi committed:
Merge remote-tracking branch 'origin/add-gradle-support'

VedJoshi committed:
Change gradle files, use Agave not Duke

VedJoshi committed:
Add JUnit tests for task list and parser

VedJoshi committed:
Merge branch 'Unit_Tests'

VedJoshi committed:
Edit build.gradle file to package as JAR file

VedJoshi committed:
Add javadoc comments for A-JavaDoc

VedJoshi committed:
Change code to adhere to coding standard

VedJoshi committed:
Add find feature to agave

VedJoshi committed:
Merge branch 'lvl9'

VedJoshi committed:
Merge branch 'Coding_Standard'

VedJoshi committed:
Fix merge conflicts, intellij is cool

george-yeo commented on src/main/java/agave/TaskList.java:
```diff
@@ -0,0 +1,124 @@
+    public void deleteTask(int taskNumber) throws AgaveException {
+        if (taskNumber > 0 && taskNumber <= tasks.size()) {
+            Task removed = tasks.remove(taskNumber - 1);
+            System.out.println("Removed: " + removed);
```
Should this method be extracted into Ui? It is possible that issues arise when the output target has to be changed.

george-yeo commented on src/main/java/agave/TaskList.java:
```diff
@@ -0,0 +1,124 @@
+     * Shows the number of tasks in the list.
+     */
+    public void showNumberOfTasks() {
+        System.out.println("Now you have " + tasks.size() + " tasks in the list.");
```
Same here for output.

george-yeo commented on src/main/java/agave/Ui.java:
```diff
@@ -0,0 +1,94 @@
+import java.util.ArrayList;
+import java.util.Scanner;
+
+public class Ui {
```
Should there be javadocs here in the Ui class and its public methods?

george-yeo commented on src/main/java/agave/Storage.java:
```diff
@@ -0,0 +1,79 @@
+     * @param tasks The list of tasks to be saved.
+     * @throws IOException If an error occurs when writing to the file.
+     */
+    public void saveTasks(ArrayList<Task> tasks) throws IOException {
```
I like the use of a BufferedWriter here to save.

george-yeo reviewed (commented):
I like how the methods are all organized and clean, and the good use of JavaDocs.

VedJoshi committed:
Bug fixes, merge did not correctly happen

VedJoshi committed:
Initial JavaFX code, time to cook

VedJoshi committed:
Update GUI and update UI class

VedJoshi committed:
More updates for GUI - fix fxml

VedJoshi committed:
Fix GUI, app runs, no user feedback

VedJoshi committed:
Refine GUI, mostly complete add CSS maybe

VedJoshi committed:
Merge branch 'Level-10'

starbucksventi commented on src/main/java/agave/Util/Ui.java:
```diff
@@ -0,0 +1,76 @@
+        makeLine();
+    }
+
+    public String searchResult(ArrayList<Task> result) {
```
Consider adding a JavaDoc comment for this and the other methods in this class. Good naming of methods and extensive use of comments elsewhere!

starbucksventi commented on src/main/java/agave/logic/ToDo.java:
```diff
@@ -0,0 +1,18 @@
+ */
+public class ToDo extends Task {
+
+    public ToDo (String description) {
```
Consider adding a JavaDoc comment to this constructor.

starbucksventi reviewed (commented):
Overall, kudos for the extensive use of OOP to delegate the functionality to specific classes and thorough use of JavaDoc comments to explain your code. Other than a few areas that are missing JavaDoc comments, looks good to merge!

VedJoshi committed:
Add assertions for task validation and idx bounds

The current task management system does not enforce input validation
throughout the codebase, relying solely on exception handling in some
cases.

Adding assertions improves the robustness and clarity of the code by
explicitly defining important assumptions about task input and
operations on task lists.

Added assertion to:
- Ensure that task descriptions are neither null nor empty when creating tasks.
- Verify that task indices are within valid bounds when accessing or modifying tasks in the task list.
- Confirm that the task number passed to marking/unmarking methods is valid.

VedJoshi committed:
Refactor main logic into CommandHandler class

The Agave class currently contains a long chain of else-if statements
for handling different commands, which makes the code harder to maintain
and read. This commit refactors the main logic into a new CommandHandler.

Changes:
- Moved command processing from Agave into CommandHandler.
- CommandHandler now handles all task commands like list, mark, unmark,
  todo, deadline, event, delete, and find.
- Agave class now delegates command execution to CommandHandler,
  resulting in cleaner, more maintainable code.

VedJoshi committed:
Merge pull request #9 from VedJoshi/branch-A-CodeQuality

Refactor main logic into CommandHandler class

VedJoshi committed:
Merge branch 'master' of https://github.com/VedJoshi/ip into branch-A-Assertions

VedJoshi committed:
Merge pull request #10 from VedJoshi/branch-A-Assertions

Branch a assertions

VedJoshi committed:
Add sort feature to sort deadlines chronologically

Implemented a new feature that allows sorting tasks, specifically
deadlines, in chronological order. This enhances the usability
of the task manager by enabling users to view their deadlines
in order of urgency.

Note: It permanently sorts the deadlines in chronological order

Changes:
- Added sortByDeadline method to TaskList to sort Deadline tasks.
- Updated CommandHandler to recognize and process the 'sort' command.
- Non-deadline tasks remain unsorted
- Updated Ui to display a message when deadlines are sorted.

VedJoshi committed:
Merge pull request #11 from VedJoshi/BCD-Extension

Add sort feature to sort deadlines chronologically

VedJoshi committed:
Refactor package naming to follow coding standard

Changed package named GUI to gui to follow
Java coding standard. Fixed incorrect braces
in CommandHandler, else statement rectified.

Minor changes for improved overall code quality.

VedJoshi committed:
Refactor FXML and move inline stles to css file

To improve maintainability, refactor FXML files to
use external CSS file for styling. Ensure a
cleaner FXML structure by removing inline styles.

Changes:
- Removed inline styles from FXML files
- Added external CSS file for styling
- Added styling as css classes for FXML elements

VedJoshi committed:
Fix dialog box resizing and add User Guide

- Added a detailed User Guide to the Agave chatbot
to help users understand the core features and
commands.
- Each feature is described with its format,
example usage, and expected output.
- Fixed a bug where the dialog box would not
resize to accommodate longer task lists

VedJoshi committed:
Rename UI.png to Ui.png

I am a silly goose and named 
it incorrectly.

VedJoshi committed:
Refactor file path handling and updated Docs

- Updated the file path for storing tasks to
dynamically use the current working directory.
- Changed file path initialization from
"./data/agave.txt" to
`System.getProperty("user.dir") +
"/data/agave.txt"`.
- Ensures consistent file storage behavior
when running the JAR file, as the `data` folder
is now always created in the directory from which
the application is executed.
- Edited the docs.

VedJoshi committed:
Merge branch 'master' of
https://github.com/VedJoshi/ip

VedJoshi committed:
Change package name to lowercase

Changed the package name from Task to
task and Util to util
